<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Spatially-dependent normalisation for spatial transcriptomics data — SpaNorm • SpaNorm</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Spatially-dependent normalisation for spatial transcriptomics data — SpaNorm"><meta name="description" content="Performs normalisation of spatial transcriptomics data using spatially-dependent spot- and gene- specific size factors."><meta property="og:description" content="Performs normalisation of spatial transcriptomics data using spatially-dependent spot- and gene- specific size factors."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SpaNorm</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/SpaNorm.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/bhuvad/SpaNorm/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Spatially-dependent normalisation for spatial transcriptomics data</h1>
      <small class="dont-index">Source: <a href="https://github.com/bhuvad/SpaNorm/blob/master/R/mainSpaNorm.R" class="external-link"><code>R/mainSpaNorm.R</code></a></small>
      <div class="d-none name"><code>SpaNorm.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Performs normalisation of spatial transcriptomics data using spatially-dependent spot- and gene- specific size factors.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">SpaNorm</span><span class="op">(</span></span>
<span>  <span class="va">spe</span>,</span>
<span>  sample.p <span class="op">=</span> <span class="fl">0.25</span>,</span>
<span>  gene.model <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nb"</span><span class="op">)</span>,</span>
<span>  adj.method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"auto"</span>, <span class="st">"logpac"</span>, <span class="st">"pearson"</span>, <span class="st">"medbio"</span>, <span class="st">"meanbio"</span><span class="op">)</span>,</span>
<span>  scale.factor <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  df.tps <span class="op">=</span> <span class="fl">6</span>,</span>
<span>  lambda.a <span class="op">=</span> <span class="fl">1e-04</span>,</span>
<span>  batch <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  tol <span class="op">=</span> <span class="fl">1e-04</span>,</span>
<span>  step.factor <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  maxit.nb <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  maxit.psi <span class="op">=</span> <span class="fl">25</span>,</span>
<span>  maxn.psi <span class="op">=</span> <span class="fl">500</span>,</span>
<span>  overwrite <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'SpatialExperiment'</span></span>
<span><span class="fu">SpaNorm</span><span class="op">(</span></span>
<span>  <span class="va">spe</span>,</span>
<span>  sample.p <span class="op">=</span> <span class="fl">0.25</span>,</span>
<span>  gene.model <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nb"</span><span class="op">)</span>,</span>
<span>  adj.method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"auto"</span>, <span class="st">"logpac"</span>, <span class="st">"pearson"</span>, <span class="st">"medbio"</span>, <span class="st">"meanbio"</span><span class="op">)</span>,</span>
<span>  scale.factor <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  df.tps <span class="op">=</span> <span class="fl">6</span>,</span>
<span>  lambda.a <span class="op">=</span> <span class="fl">1e-04</span>,</span>
<span>  batch <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  tol <span class="op">=</span> <span class="fl">1e-04</span>,</span>
<span>  step.factor <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  maxit.nb <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  maxit.psi <span class="op">=</span> <span class="fl">25</span>,</span>
<span>  maxn.psi <span class="op">=</span> <span class="fl">500</span>,</span>
<span>  overwrite <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-spe">spe<a class="anchor" aria-label="anchor" href="#arg-spe"></a></dt>
<dd><p>a SpatialExperiment or Seurat object, with the count data stored in 'counts' or 'data' assays respectively.</p></dd>


<dt id="arg-sample-p">sample.p<a class="anchor" aria-label="anchor" href="#arg-sample-p"></a></dt>
<dd><p>a numeric, specifying the (maximum) proportion of cells/spots to sample for model fitting (default is 0.25).</p></dd>


<dt id="arg-gene-model">gene.model<a class="anchor" aria-label="anchor" href="#arg-gene-model"></a></dt>
<dd><p>a character, specifying the model to use for gene/protein abundances (default 'nb'). This should be 'nb' for count based datasets.</p></dd>


<dt id="arg-adj-method">adj.method<a class="anchor" aria-label="anchor" href="#arg-adj-method"></a></dt>
<dd><p>a character, specifying the method to use to adjust the data (default 'auto', see details)</p></dd>


<dt id="arg-scale-factor">scale.factor<a class="anchor" aria-label="anchor" href="#arg-scale-factor"></a></dt>
<dd><p>a numeric, specifying the sample-specific scaling factor to scale the adjusted count.</p></dd>


<dt id="arg-df-tps">df.tps<a class="anchor" aria-label="anchor" href="#arg-df-tps"></a></dt>
<dd><p>a numeric, specifying the degrees of freedom for the thin-plate spline (default is 6).</p></dd>


<dt id="arg-lambda-a">lambda.a<a class="anchor" aria-label="anchor" href="#arg-lambda-a"></a></dt>
<dd><p>a numeric, specifying the smoothing parameter for regularizing regression coefficients (default is 0.0001). Actual lambda.a used is lambda.a * ncol(spe).</p></dd>


<dt id="arg-batch">batch<a class="anchor" aria-label="anchor" href="#arg-batch"></a></dt>
<dd><p>a vector or numeric matrix, specifying the batch design to regress out (default NULL, representing no batch effects). See details for more information on how to define this variable.</p></dd>


<dt id="arg-tol">tol<a class="anchor" aria-label="anchor" href="#arg-tol"></a></dt>
<dd><p>a numeric, specifying the tolerance for convergence (default is 1e-4).</p></dd>


<dt id="arg-step-factor">step.factor<a class="anchor" aria-label="anchor" href="#arg-step-factor"></a></dt>
<dd><p>a numeric, specifying the multiplicative factor to decrease IRLS step by when log-likelihood diverges (default is 0.5).</p></dd>


<dt id="arg-maxit-nb">maxit.nb<a class="anchor" aria-label="anchor" href="#arg-maxit-nb"></a></dt>
<dd><p>a numeric, specifying the maximum number of IRLS iteration for estimating NB mean parameters for a given dispersion parameter (default is 50).</p></dd>


<dt id="arg-maxit-psi">maxit.psi<a class="anchor" aria-label="anchor" href="#arg-maxit-psi"></a></dt>
<dd><p>a numeric, specifying the maximum number of IRLS iterations to estimate the dispersion parameter (default is 25).</p></dd>


<dt id="arg-maxn-psi">maxn.psi<a class="anchor" aria-label="anchor" href="#arg-maxn-psi"></a></dt>
<dd><p>a numeric, specifying the maximum number of cells/spots to sample for dispersion estimation (default is 500).</p></dd>


<dt id="arg-overwrite">overwrite<a class="anchor" aria-label="anchor" href="#arg-overwrite"></a></dt>
<dd><p>a logical, specifying whether to force recomputation and overwrite an existing fit (default FALSE). Note that if df.tps, batch, lambda.a, or gene.model are changed, the model is recomputed and overwritten.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>a logical, specifying whether to show update messages (default TRUE).</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>other parameters fitting parameters.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>a SpatialExperiment or Seurat object with the adjusted data stored in 'logcounts' or 'data', respectively.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>SpaNorm works by first fitting a spatial regression model for library size to the data. Normalised data can then be computed using various adjustment approaches. When a negative binomial gene-model is used, the data can be adjusted using the following approaches: 'logpac', 'pearson', 'medbio', and 'meanbio'.</p>
<p>Batch effects can be specified using the <code>batch</code> parameter. If this parameter is a vector, a design matrix will be created within the function using <code>model.matrix</code>. If a custom design is provided in the form of a numeric matrix, this should ideally be created using <code>model.matrix</code>. The batch matrix should be created with an intercept term. The SpaNorm function will automatically detect the intercept term and remove the relevant column. Alternatively, users can subset the model matrix to remove this column manually. Please note that the model formula should include the intercept term and that the intercept column should be subset out after.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">HumanDLPFC</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># \donttest{</span></span></span>
<span class="r-in"><span><span class="fu">SpaNorm</span><span class="op">(</span><span class="va">HumanDLPFC</span>, sample.p <span class="op">=</span> <span class="fl">0.05</span>, df.tps <span class="op">=</span> <span class="fl">2</span>, tol <span class="op">=</span> <span class="fl">1e-2</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: SpatialExperiment</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: SingleCellExperiment</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: SummarizedExperiment</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: MatrixGenerics</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: matrixStats</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘MatrixGenerics’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:matrixStats’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colCounts, colCummaxs, colCummins, colCumprods, colCumsums,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colWeightedMeans, colWeightedMedians, colWeightedSds,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowWeightedMads, rowWeightedMeans, rowWeightedMedians,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowWeightedSds, rowWeightedVars</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: GenomicRanges</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: stats4</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: BiocGenerics</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘BiocGenerics’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:stats’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     IQR, mad, sd, var, xtabs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:base’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     Filter, Find, Map, Position, Reduce, anyDuplicated, aperm, append,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     as.data.frame, basename, cbind, colnames, dirname, do.call,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     duplicated, eval, evalq, get, grep, grepl, intersect, is.unsorted,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     lapply, mapply, match, mget, order, paste, pmax, pmax.int, pmin,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     pmin.int, rank, rbind, rownames, sapply, setdiff, table, tapply,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     union, unique, unsplit, which.max, which.min</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: S4Vectors</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘S4Vectors’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following object is masked from ‘package:utils’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     findMatches</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:base’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     I, expand.grid, unname</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: IRanges</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: GenomeInfoDb</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: Biobase</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Welcome to Bioconductor</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     Vignettes contain introductory material; view with</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     'browseVignettes()'. To cite Bioconductor, see</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     'citation("Biobase")', and for packages 'citation("pkgname")'.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘Biobase’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following object is masked from ‘package:MatrixGenerics’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowMedians</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:matrixStats’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     anyMissing, rowMedians</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>replacing previous import ‘S4Arrays::makeNindexFromArrayViewport’ by ‘DelayedArray::makeNindexFromArrayViewport’ when loading ‘SummarizedExperiment’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> (1/2) Fitting SpaNorm model</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 201 cells/spots sampled to fit model</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> iter:  1, estimating gene-wise dispersion</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> iter:  1, log-likelihood: -1131632.440359</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> iter:  1, fitting NB model</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> iter:  1, iter:  1, log-likelihood: -1131632.440359</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> iter:  1, iter:  2, log-likelihood: -813184.377314</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> iter:  1, iter:  3, log-likelihood: -726819.709782</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> iter:  1, iter:  4, log-likelihood: -708014.945154</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> iter:  1, iter:  5, log-likelihood: -704554.727696</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> iter:  1, iter:  6, log-likelihood: -703928.449409</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> iter:  1, iter:  7, log-likelihood: -703796.923679</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> iter:  1, iter:  8, log-likelihood: -703757.707708 (converged)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> iter:  2, estimating gene-wise dispersion</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> iter:  2, log-likelihood: -703309.607508</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> iter:  2, fitting NB model</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> iter:  2, iter:  1, log-likelihood: -703309.607508</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> iter:  2, iter:  2, log-likelihood: -703040.415059</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> iter:  2, iter:  3, log-likelihood: -703023.823163 (converged)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> iter:  3, log-likelihood: -703023.823163 (converged)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> (2/2) Normalising data</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> class: SpatialExperiment </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> dim: 5076 4015 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> metadata(1): SpaNorm</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> assays(2): counts logcounts</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rownames(5076): ENSG00000188976 ENSG00000188290 ... ENSG00000198727</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ENSG00000278817</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rowData names(2): gene_name gene_biotype</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colnames(4015): AAACAAGTATCTCCCA-1 AAACACCAATAACTGC-1 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   TTGTTTCCATACAACT-1 TTGTTTGTGTAAATTC-1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colData names(3): cell_count sample_id AnnotatedCluster</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> reducedDimNames(0):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> mainExpName: NULL</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> altExpNames(0):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> spatialCoords names(2) : pxl_col_in_fullres pxl_row_in_fullres</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> imgData names(4): sample_id image_id data scaleFactor</span>
<span class="r-in"><span><span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Dharmesh D. Bhuva, Agus Salim, Ahmed Mohamed.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

