<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>SpaNorm: Spatially aware library size normalisation ‚Ä¢ SpaNorm</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="SpaNorm: Spatially aware library size normalisation">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SpaNorm</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/SpaNorm.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/bhuvad/SpaNorm/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>SpaNorm: Spatially aware library size normalisation</h1>
                        <h4 data-toc-skip class="author">Dharmesh D.
Bhuva and Agus Salim</h4>
            
            <h4 data-toc-skip class="date">9 February 2025</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/bhuvad/SpaNorm/blob/master/vignettes/SpaNorm.Rmd" class="external-link"><code>vignettes/SpaNorm.Rmd</code></a></small>
      <div class="d-none name"><code>SpaNorm.Rmd</code></div>
    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>This package implements the spatially aware library size
      normalisation algorithm, SpaNorm. SpaNorm normalises out library
      size effects while retaining biology through the modelling of
      smooth functions for each effect. Normalisation is performed in a
      gene- and cell-/spot- specific manner, yielding library size
      adjusted data.</p>
    </div>
    
<div class="section level2">
<h2 id="spanorm">SpaNorm<a class="anchor" aria-label="anchor" href="#spanorm"></a>
</h2>
<p>SpaNorm is a spatially aware library size normalisation method that
removes library size effects, while retaining biology. Library sizes
need to be removed from molecular datasets to allow comparisons across
observations, in this case, across space. Bhuva et al. <span class="citation">(Bhuva et al. 2024)</span> and Atta et al. <span class="citation">(Atta et al. 2024)</span> have shown that standard
single-cell inspired library size normalisation approaches are not
appropriate for spatial molecular datasets as they often remove
biological signals while doing so. This is because library size
confounds biology in spatial molecular data.</p>
<div class="float">
<img src="SpaNormWorkflow.png" alt="The SpaNorm workflow: SpaNorm takes the gene expression data and spatial coordinates as inputs. Using a gene-wise model (e.g., Negative Binomial (NB)), SpaNorm decomposes spatially-smooth variation into those unrelated to library size (LS), representing the underlying true biology and those related to library size. The adjusted data is then produced by keeping only the variation unrelated to library size."><div class="figcaption"><em>The SpaNorm workflow: SpaNorm takes the gene
expression data and spatial coordinates as inputs. Using a gene-wise
model (e.g., Negative Binomial (NB)), SpaNorm decomposes
spatially-smooth variation into those unrelated to library size (LS),
representing the underlying true biology and those related to library
size. The adjusted data is then produced by keeping only the variation
unrelated to library size.</em></div>
</div>
<p>SpaNorm uses a unique approach to spatially constraint modelling
approach to model gene expression (e.g., counts) and remove library size
effects, while retaining biology. It achieves this through three key
innovations:</p>
<ol style="list-style-type: decimal">
<li>Optmial decomposition of spatial variation into spatially smooth
library size associated (technical) and library size independent
(biology) variation using generalized linear models (GLMs).</li>
<li>Computing spatially smooth functions (using thin plate splines) to
represent the gene- and location-/cell-/spot- specific size
factors.</li>
<li>Adjustment of data using percentile adjusted counts (PAC) <span class="citation">(Salim et al. 2022)</span>, as well as other adjustment
approaches (e.g., Pearson).</li>
</ol>
<p>The SpaNorm package can be installed as follows:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># release version</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"SpaNorm"</span><span class="op">)</span></span>
<span><span class="co"># development version from GitHub</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"bhuvad/SpaNorm"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="load-count-data">Load count data<a class="anchor" aria-label="anchor" href="#load-count-data"></a>
</h2>
<p>We begin by loading some example 10x Visium data profiling the
dorsolateral prefrontal cortex (DLPFC) of the human brain. The data has
~4,000 spots and covers genome-wide measurements. The example data here
is filtered to remove lowly expressed genes (using
<code>filterGenes(HumanDLPFC, prop = 0.1)</code>). This filtering
retains genes that are expressed in at least 10% of cells.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bhuvad.github.io/SpaNorm" class="external-link">SpaNorm</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/drighelli/SpatialExperiment" class="external-link">SpatialExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># load sample data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">HumanDLPFC</span><span class="op">)</span></span>
<span><span class="co"># change gene IDs to gene names</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">HumanDLPFC</span><span class="op">)</span> <span class="op">=</span> <span class="fu">rowData</span><span class="op">(</span><span class="va">HumanDLPFC</span><span class="op">)</span><span class="op">$</span><span class="va">gene_name</span></span>
<span><span class="va">HumanDLPFC</span></span>
<span><span class="co">#&gt; class: SpatialExperiment </span></span>
<span><span class="co">#&gt; dim: 5076 4015 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(5076): NOC2L HES4 ... MT-CYB AC007325.4</span></span>
<span><span class="co">#&gt; rowData names(2): gene_name gene_biotype</span></span>
<span><span class="co">#&gt; colnames(4015): AAACAAGTATCTCCCA-1 AAACACCAATAACTGC-1 ...</span></span>
<span><span class="co">#&gt;   TTGTTTCCATACAACT-1 TTGTTTGTGTAAATTC-1</span></span>
<span><span class="co">#&gt; colData names(3): cell_count sample_id AnnotatedCluster</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : pxl_col_in_fullres pxl_row_in_fullres</span></span>
<span><span class="co">#&gt; imgData names(4): sample_id image_id data scaleFactor</span></span>
<span></span>
<span><span class="co"># plot regions</span></span>
<span><span class="va">p_region</span> <span class="op">=</span> <span class="fu"><a href="../reference/plotSpatial.html">plotSpatial</a></span><span class="op">(</span><span class="va">HumanDLPFC</span>, colour <span class="op">=</span> <span class="va">AnnotatedCluster</span>, size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_colour_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Paired"</span>, guide <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>override.aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>shape <span class="op">=</span> <span class="fl">15</span>, size <span class="op">=</span> <span class="fl">5</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Region"</span><span class="op">)</span></span>
<span><span class="va">p_region</span></span></code></pre></div>
<p><img src="SpaNorm_files/figure-html/unnamed-chunk-3-1.png" width="384"></p>
<p>The <code>filterGenes</code> function returns a logical vector
indicating which genes should be kept.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># filter genes expressed in 20% of spots</span></span>
<span><span class="va">keep</span> <span class="op">=</span> <span class="fu"><a href="../reference/filterGenes.html">filterGenes</a></span><span class="op">(</span><span class="va">HumanDLPFC</span>, <span class="fl">0.2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">keep</span><span class="op">)</span></span>
<span><span class="co">#&gt; keep</span></span>
<span><span class="co">#&gt; FALSE  TRUE </span></span>
<span><span class="co">#&gt;  2568  2508</span></span>
<span><span class="co"># subset genes</span></span>
<span><span class="va">HumanDLPFC</span> <span class="op">=</span> <span class="va">HumanDLPFC</span><span class="op">[</span><span class="va">keep</span>, <span class="op">]</span></span></code></pre></div>
<p>The log-transformed raw counts are visualised below for the gene
<em>MOBP</em> which is a marker of oligodendrocytes enriched in the
white matter (WM) <span class="citation">(Maynard et al. 2021)</span>.
Despite being a marker of this region, we see that it is in fact absent
from the white matter region.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">logcounts</a></span><span class="op">(</span><span class="va">HumanDLPFC</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">HumanDLPFC</span><span class="op">)</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p_counts</span> <span class="op">=</span> <span class="fu"><a href="../reference/plotSpatial.html">plotSpatial</a></span><span class="op">(</span></span>
<span>    <span class="va">HumanDLPFC</span>,</span>
<span>    colour <span class="op">=</span> <span class="va">MOBP</span>,</span>
<span>    what <span class="op">=</span> <span class="st">"expression"</span>,</span>
<span>    assay <span class="op">=</span> <span class="st">"logcounts"</span>,</span>
<span>    size <span class="op">=</span> <span class="fl">0.5</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_colour_viridis_c</a></span><span class="op">(</span>option <span class="op">=</span> <span class="st">"F"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"logCounts"</span><span class="op">)</span></span>
<span><span class="va">p_region</span> <span class="op">+</span> <span class="va">p_counts</span></span></code></pre></div>
<p><img src="SpaNorm_files/figure-html/unnamed-chunk-5-1.png" width="720"></p>
</div>
<div class="section level2">
<h2 id="normalise-count-data">Normalise count data<a class="anchor" aria-label="anchor" href="#normalise-count-data"></a>
</h2>
<p>SpaNorm normalises data in two steps: (1) fitting the SpaNorm model
of library sizes; (2) adjusting data using the fit model. A single call
to the <code><a href="../reference/SpaNorm.html">SpaNorm()</a></code> function is enough to run these two steps.
To speed up computation, the model is fit using a smaller proportion of
spots/cells (default is 0.25). The can be modified using the
<code>sample.p</code> parameter.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">36</span><span class="op">)</span></span>
<span><span class="va">HumanDLPFC</span> <span class="op">=</span> <span class="fu"><a href="../reference/SpaNorm.html">SpaNorm</a></span><span class="op">(</span><span class="va">HumanDLPFC</span><span class="op">)</span></span>
<span><span class="co">#&gt; (1/2) Fitting SpaNorm model</span></span>
<span><span class="co">#&gt; 1004 cells/spots sampled to fit model</span></span>
<span><span class="co">#&gt; iter:  1, estimating gene-wise dispersion</span></span>
<span><span class="co">#&gt; iter:  1, log-likelihood: -3185253.033796</span></span>
<span><span class="co">#&gt; iter:  1, fitting NB model</span></span>
<span><span class="co">#&gt; iter:  1, iter:  1, log-likelihood: -3185253.033796</span></span>
<span><span class="co">#&gt; iter:  1, iter:  2, log-likelihood: -2567793.267240</span></span>
<span><span class="co">#&gt; iter:  1, iter:  3, log-likelihood: -2450249.643552</span></span>
<span><span class="co">#&gt; iter:  1, iter:  4, log-likelihood: -2414033.201817</span></span>
<span><span class="co">#&gt; iter:  1, iter:  5, log-likelihood: -2398027.810891</span></span>
<span><span class="co">#&gt; iter:  1, iter:  6, log-likelihood: -2392561.924421</span></span>
<span><span class="co">#&gt; iter:  1, iter:  7, log-likelihood: -2390638.768170</span></span>
<span><span class="co">#&gt; iter:  1, iter:  8, log-likelihood: -2389892.874670</span></span>
<span><span class="co">#&gt; iter:  1, iter:  9, log-likelihood: -2389530.867843</span></span>
<span><span class="co">#&gt; iter:  1, iter: 10, log-likelihood: -2389383.012576 (converged)</span></span>
<span><span class="co">#&gt; iter:  2, estimating gene-wise dispersion</span></span>
<span><span class="co">#&gt; iter:  2, log-likelihood: -2384794.935592</span></span>
<span><span class="co">#&gt; iter:  2, fitting NB model</span></span>
<span><span class="co">#&gt; iter:  2, iter:  1, log-likelihood: -2384794.935592</span></span>
<span><span class="co">#&gt; iter:  2, iter:  2, log-likelihood: -2383118.209922</span></span>
<span><span class="co">#&gt; iter:  2, iter:  2, log-likelihood: -2383118.209922</span></span>
<span><span class="co">#&gt; iter:  2, iter:  2, log-likelihood: -2383118.209922</span></span>
<span><span class="co">#&gt; iter:  2, iter:  3, log-likelihood: -2383118.209922 (converged)</span></span>
<span><span class="co">#&gt; iter:  3, estimating gene-wise dispersion</span></span>
<span><span class="co">#&gt; iter:  3, log-likelihood: -2383101.131211</span></span>
<span><span class="co">#&gt; iter:  3, fitting NB model</span></span>
<span><span class="co">#&gt; iter:  3, iter:  1, log-likelihood: -2383101.131211</span></span>
<span><span class="co">#&gt; iter:  3, iter:  2, log-likelihood: -2383083.477463</span></span>
<span><span class="co">#&gt; iter:  3, iter:  3, log-likelihood: -2383049.505902 (converged)</span></span>
<span><span class="co">#&gt; iter:  4, log-likelihood: -2383049.505902 (converged)</span></span>
<span><span class="co">#&gt; (2/2) Normalising data</span></span>
<span><span class="va">HumanDLPFC</span></span>
<span><span class="co">#&gt; class: SpatialExperiment </span></span>
<span><span class="co">#&gt; dim: 2508 4015 </span></span>
<span><span class="co">#&gt; metadata(1): SpaNorm</span></span>
<span><span class="co">#&gt; assays(2): counts logcounts</span></span>
<span><span class="co">#&gt; rownames(2508): ISG15 SDF4 ... MT-ND6 MT-CYB</span></span>
<span><span class="co">#&gt; rowData names(2): gene_name gene_biotype</span></span>
<span><span class="co">#&gt; colnames(4015): AAACAAGTATCTCCCA-1 AAACACCAATAACTGC-1 ...</span></span>
<span><span class="co">#&gt;   TTGTTTCCATACAACT-1 TTGTTTGTGTAAATTC-1</span></span>
<span><span class="co">#&gt; colData names(3): cell_count sample_id AnnotatedCluster</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : pxl_col_in_fullres pxl_row_in_fullres</span></span>
<span><span class="co">#&gt; imgData names(4): sample_id image_id data scaleFactor</span></span></code></pre></div>
<p>The above output (which can be switched off by setting
<code>verbose = FALSE</code>), shows the two steps of normalisation. In
the model fitting step, 1004 cells/spots are used to fit the negative
binomial (NB) model. Subsequent output shows that this fit is performed
by alternating between estimation of the dispersion parameter and
estimation of the NB parameters by fixing the dispersion. The output
also shows that each intermediate fit converges, and so does the final
fit. The accuracy of the fit can be controlled by modifying the
tolerance parameter <code>tol</code> (default <code>1e-4</code>).</p>
<p>Next, data is adjusted using the fit model. The following approaches
are implemented for count data:</p>
<ol style="list-style-type: decimal">
<li>
<code>adj.method = "logpac"</code> (default) - percentile adjusted
counts (PAC) which estimates the count for each gene at each
location/spot/cell using a model that does not contain unwanted effects
such as the library size.</li>
<li>
<code>adj.method = "person"</code> - Pearson residuals from
factoring out unwanted effects.</li>
<li>
<code>adj.method = "meanbio"</code> - the mean of each gene at each
location estimated from the biological component of the model.</li>
<li>
<code>adj.method = "medbio"</code> - the median of each gene at each
location estimated from the biological component of the model.</li>
</ol>
<p>These data are stored in the <code>logcounts</code> assay of the
SpatialExperiment object. After normalisation, we see that MOBP is
enriched in the white matter.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p_logpac</span> <span class="op">=</span> <span class="fu"><a href="../reference/plotSpatial.html">plotSpatial</a></span><span class="op">(</span></span>
<span>    <span class="va">HumanDLPFC</span>,</span>
<span>    colour <span class="op">=</span> <span class="va">MOBP</span>,</span>
<span>    what <span class="op">=</span> <span class="st">"expression"</span>,</span>
<span>    assay <span class="op">=</span> <span class="st">"logcounts"</span>,</span>
<span>    size <span class="op">=</span> <span class="fl">0.5</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_colour_viridis_c</a></span><span class="op">(</span>option <span class="op">=</span> <span class="st">"F"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"logPAC"</span><span class="op">)</span></span>
<span><span class="va">p_region</span> <span class="op">+</span> <span class="va">p_logpac</span></span></code></pre></div>
<p><img src="SpaNorm_files/figure-html/unnamed-chunk-7-1.png" width="720"></p>
</div>
<div class="section level2">
<h2 id="computing-alternative-adjustments-using-a-precomputed-spanorm-fit">Computing alternative adjustments using a precomputed SpaNorm
fit<a class="anchor" aria-label="anchor" href="#computing-alternative-adjustments-using-a-precomputed-spanorm-fit"></a>
</h2>
<p>As no appropriate slot exists for storing model parameters, we
currently save them in the metadata slot with the name ‚ÄúSpaNorm‚Äù. This
also means that subsetting features (i.e., genes) or observations (i.e.,
cells/spots/loci) does not subset the model. In such an instance, the
SpaNorm function will realise that the model no longer matches the data
and re-estimates when called. If instead the model is valid for the
data, the existing fit is extracted and reused.</p>
<p>The fit can be manually retrieved as below for users wishing to reuse
the model outside the SpaNorm framework. Otherwise, calling
<code><a href="../reference/SpaNorm.html">SpaNorm()</a></code> on an object containing the fit will
automatically use it.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># manually retrieve model</span></span>
<span><span class="va">fit.spanorm</span> <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">HumanDLPFC</span><span class="op">)</span><span class="op">$</span><span class="va">SpaNorm</span></span>
<span><span class="va">fit.spanorm</span></span>
<span><span class="co">#&gt; SpaNormFit</span></span>
<span><span class="co">#&gt; Data: 2508 genes, 4015 cells/spots</span></span>
<span><span class="co">#&gt; Gene model: nb</span></span>
<span><span class="co">#&gt; Degrees of freedom for TPS (y,x): Biology (6,6), LS (6,6)</span></span>
<span><span class="co">#&gt; Spots/cells sampled: 25%</span></span>
<span><span class="co">#&gt; Regularisation parameter: 1e-04</span></span>
<span><span class="co">#&gt; Batch:  NULL</span></span>
<span><span class="co">#&gt; log-likelihood (per-iteration):  num [1:3] -2389383 -2383118 -2383050</span></span>
<span><span class="co">#&gt; W:  num [1:4015, 1:73] 0.2645 0.4736 0.0547 -0.1756 0.6039 ...</span></span>
<span><span class="co">#&gt; W:  - attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">#&gt; W:   ..$ : chr [1:4015] "1" "2" "3" "4" ...</span></span>
<span><span class="co">#&gt; W:   ..$ : chr [1:73] "logLS" "bs.xy1" "bs.xy2" "bs.xy3" ...</span></span>
<span><span class="co">#&gt; alpha:  num [1:2508, 1:73] 1.01 1.01 1.01 1.01 1.01 ...</span></span>
<span><span class="co">#&gt; gmean:  num [1:2508] -1.249 -1.174 -1.266 -1.436 -0.386 ...</span></span>
<span><span class="co">#&gt; psi:  num [1:2508] 9.77e-05 9.77e-05 9.77e-05 9.77e-05 1.52e-02 ...</span></span>
<span><span class="co">#&gt; wtype:  Factor w/ 3 levels "batch","biology",..: 3 2 2 2 2 2 2 2 2 2 ...</span></span>
<span><span class="co">#&gt; sampling: all (4015), glm (1004), dispersion (500)</span></span></code></pre></div>
<p>When a valid fit exists in the object, only the adjustment step is
performed. The model is recomputed if <code>overwrite = TRUE</code> or
any of the following parameters change: degrees of freedom
(<code>df.tps</code>), penalty parameters(<code>lambda.a</code>), object
dimensions, or <code>batch</code> specification. Alternative adjustments
can be computed as below and stored to the <code>logcounts</code>
assay.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Pearson residuals</span></span>
<span><span class="va">HumanDLPFC</span> <span class="op">=</span> <span class="fu"><a href="../reference/SpaNorm.html">SpaNorm</a></span><span class="op">(</span><span class="va">HumanDLPFC</span>, adj.method <span class="op">=</span> <span class="st">"pearson"</span><span class="op">)</span></span>
<span><span class="va">p_pearson</span> <span class="op">=</span> <span class="fu"><a href="../reference/plotSpatial.html">plotSpatial</a></span><span class="op">(</span></span>
<span>    <span class="va">HumanDLPFC</span>,</span>
<span>    colour <span class="op">=</span> <span class="va">MOBP</span>,</span>
<span>    what <span class="op">=</span> <span class="st">"expression"</span>,</span>
<span>    assay <span class="op">=</span> <span class="st">"logcounts"</span>,</span>
<span>    size <span class="op">=</span> <span class="fl">0.5</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_colour_viridis_c</a></span><span class="op">(</span>option <span class="op">=</span> <span class="st">"F"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Pearson"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># meanbio residuals</span></span>
<span><span class="va">HumanDLPFC</span> <span class="op">=</span> <span class="fu"><a href="../reference/SpaNorm.html">SpaNorm</a></span><span class="op">(</span><span class="va">HumanDLPFC</span>, adj.method <span class="op">=</span> <span class="st">"meanbio"</span><span class="op">)</span></span>
<span><span class="va">p_meanbio</span> <span class="op">=</span> <span class="fu"><a href="../reference/plotSpatial.html">plotSpatial</a></span><span class="op">(</span></span>
<span>    <span class="va">HumanDLPFC</span>,</span>
<span>    colour <span class="op">=</span> <span class="va">MOBP</span>,</span>
<span>    what <span class="op">=</span> <span class="st">"expression"</span>,</span>
<span>    assay <span class="op">=</span> <span class="st">"logcounts"</span>,</span>
<span>    size <span class="op">=</span> <span class="fl">0.5</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_colour_viridis_c</a></span><span class="op">(</span>option <span class="op">=</span> <span class="st">"F"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Mean biology"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># meanbio residuals</span></span>
<span><span class="va">HumanDLPFC</span> <span class="op">=</span> <span class="fu"><a href="../reference/SpaNorm.html">SpaNorm</a></span><span class="op">(</span><span class="va">HumanDLPFC</span>, adj.method <span class="op">=</span> <span class="st">"medbio"</span><span class="op">)</span></span>
<span><span class="va">p_medbio</span> <span class="op">=</span> <span class="fu"><a href="../reference/plotSpatial.html">plotSpatial</a></span><span class="op">(</span></span>
<span>    <span class="va">HumanDLPFC</span>,</span>
<span>    colour <span class="op">=</span> <span class="va">MOBP</span>,</span>
<span>    what <span class="op">=</span> <span class="st">"expression"</span>,</span>
<span>    assay <span class="op">=</span> <span class="st">"logcounts"</span>,</span>
<span>    size <span class="op">=</span> <span class="fl">0.5</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_colour_viridis_c</a></span><span class="op">(</span>option <span class="op">=</span> <span class="st">"F"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Median biology"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p_region</span> <span class="op">+</span> <span class="va">p_counts</span> <span class="op">+</span> <span class="va">p_logpac</span> <span class="op">+</span> <span class="va">p_pearson</span> <span class="op">+</span> <span class="va">p_meanbio</span> <span class="op">+</span> <span class="va">p_medbio</span> <span class="op">+</span> <span class="fu">plot_layout</span><span class="op">(</span>ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p><img src="SpaNorm_files/figure-html/unnamed-chunk-9-1.png" width="1104"></p>
<p>The mean biology adjustment shows a significant enrichment of the
<em>MOBP</em> gene in the white matter. As the overall counts of this
gene are low in this sample, other methods show less discriminative
power.</p>
</div>
<div class="section level2">
<h2 id="varying-model-complexity">Varying model complexity<a class="anchor" aria-label="anchor" href="#varying-model-complexity"></a>
</h2>
<p>The complexity of the spatial smoothing function is determined by the
<code>df.tps</code> parameter where larger values result in more
complicated functions (default 6).</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># df.tps = 2</span></span>
<span><span class="va">HumanDLPFC_df2</span> <span class="op">=</span> <span class="fu"><a href="../reference/SpaNorm.html">SpaNorm</a></span><span class="op">(</span><span class="va">HumanDLPFC</span>, df.tps <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">p_logpac_2</span> <span class="op">=</span> <span class="fu"><a href="../reference/plotSpatial.html">plotSpatial</a></span><span class="op">(</span></span>
<span>    <span class="va">HumanDLPFC</span>,</span>
<span>    colour <span class="op">=</span> <span class="va">MOBP</span>,</span>
<span>    what <span class="op">=</span> <span class="st">"expression"</span>,</span>
<span>    assay <span class="op">=</span> <span class="st">"logcounts"</span>,</span>
<span>    size <span class="op">=</span> <span class="fl">0.5</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_colour_viridis_c</a></span><span class="op">(</span>option <span class="op">=</span> <span class="st">"F"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"logPAC (df.tps = 2)"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># df.tps = 6 (default)</span></span>
<span><span class="va">p_logpac_6</span> <span class="op">=</span> <span class="va">p_logpac</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"logPAC (df.tps = 6)"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p_logpac_2</span> <span class="op">+</span> <span class="va">p_logpac_6</span></span></code></pre></div>
<p><img src="SpaNorm_files/figure-html/unnamed-chunk-10-1.png" width="720"></p>
</div>
<div class="section level2">
<h2 id="enhancing-signal">Enhancing signal<a class="anchor" aria-label="anchor" href="#enhancing-signal"></a>
</h2>
<p>As the counts for the MOBP gene are very low, we see artifacts in the
adjusted counts. As we have a model for the genes, we can increase the
signal by adjusting all means by a constant factor. Applying a scale
factor of 4 shows how the adjusted data are more continuous, with
significant enrichment in the white matter.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># scale.factor = 1 (default)</span></span>
<span><span class="va">HumanDLPFC</span> <span class="op">=</span> <span class="fu"><a href="../reference/SpaNorm.html">SpaNorm</a></span><span class="op">(</span><span class="va">HumanDLPFC</span>, scale.factor <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">p_logpac_sf1</span> <span class="op">=</span> <span class="fu"><a href="../reference/plotSpatial.html">plotSpatial</a></span><span class="op">(</span></span>
<span>    <span class="va">HumanDLPFC</span>,</span>
<span>    colour <span class="op">=</span> <span class="va">MOBP</span>,</span>
<span>    what <span class="op">=</span> <span class="st">"expression"</span>,</span>
<span>    assay <span class="op">=</span> <span class="st">"logcounts"</span>,</span>
<span>    size <span class="op">=</span> <span class="fl">0.5</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_colour_viridis_c</a></span><span class="op">(</span>option <span class="op">=</span> <span class="st">"F"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"logPAC (scale.factor = 1)"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># scale.factor = 4</span></span>
<span><span class="va">HumanDLPFC</span> <span class="op">=</span> <span class="fu"><a href="../reference/SpaNorm.html">SpaNorm</a></span><span class="op">(</span><span class="va">HumanDLPFC</span>, scale.factor <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span>
<span><span class="va">p_logpac_sf4</span> <span class="op">=</span> <span class="fu"><a href="../reference/plotSpatial.html">plotSpatial</a></span><span class="op">(</span></span>
<span>    <span class="va">HumanDLPFC</span>,</span>
<span>    colour <span class="op">=</span> <span class="va">MOBP</span>,</span>
<span>    what <span class="op">=</span> <span class="st">"expression"</span>,</span>
<span>    assay <span class="op">=</span> <span class="st">"logcounts"</span>,</span>
<span>    size <span class="op">=</span> <span class="fl">0.5</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_colour_viridis_c</a></span><span class="op">(</span>option <span class="op">=</span> <span class="st">"F"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"logPAC (scale.factor = 4)"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p_logpac_sf1</span> <span class="op">+</span> <span class="va">p_logpac_sf4</span> <span class="op">+</span> <span class="fu">plot_layout</span><span class="op">(</span>ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="SpaNorm_files/figure-html/unnamed-chunk-11-1.png" width="720"></p>
</div>
<div class="section level2">
<h2 id="exploring-learnt-functions">Exploring learnt functions<a class="anchor" aria-label="anchor" href="#exploring-learnt-functions"></a>
</h2>
<p>The <code><a href="../reference/plotCovariate.html">plotCovariate()</a></code> function can be used to explore the
learnt functions. We could study what the model has learnt about the
biology and library size effects of the <em>MOBP</em> gene.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span> <span class="op">=</span> <span class="fu"><a href="../reference/plotCovariate.html">plotCovariate</a></span><span class="op">(</span><span class="va">HumanDLPFC</span>, colour <span class="op">=</span> <span class="va">MOBP</span>, covariate <span class="op">=</span> <span class="st">"biology"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_colour_viridis_c</a></span><span class="op">(</span>option <span class="op">=</span> <span class="st">"F"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Biology"</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">=</span> <span class="fu"><a href="../reference/plotCovariate.html">plotCovariate</a></span><span class="op">(</span><span class="va">HumanDLPFC</span>, colour <span class="op">=</span> <span class="va">MOBP</span>, covariate <span class="op">=</span> <span class="st">"ls"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_colour_viridis_c</a></span><span class="op">(</span>option <span class="op">=</span> <span class="st">"F"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Library size effect"</span><span class="op">)</span></span>
<span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span></span></code></pre></div>
<p><img src="SpaNorm_files/figure-html/unnamed-chunk-12-1.png" width="720"></p>
</div>
<div class="section level2">
<h2 id="identifying-spatially-variable-genes">Identifying spatially variable genes<a class="anchor" aria-label="anchor" href="#identifying-spatially-variable-genes"></a>
</h2>
<p>The <code><a href="../reference/SpaNormSVG.html">SpaNormSVG()</a></code> function can be used to identify
spatially variable genes (SVGs) in the data. This function fits a nested
model without the biological function of the form:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>l</mi><mi>o</mi><mi>g</mi><msub><mi>Œº</mi><mrow><mi>c</mi><mo>,</mo><mi>g</mi></mrow></msub><mo>=</mo><msub><mo>ùúç</mo><mi>g</mi></msub><mo>+</mo><mo stretchy="false" form="prefix">{</mo><mi>Œ±</mi><mo>+</mo><msub><mi>h</mi><mi>g</mi></msub><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>x</mi><mi>c</mi></msub><mo>,</mo><msub><mi>y</mi><mi>c</mi></msub><mo>;</mo><msub><mi>Œ≥</mi><mi>g</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow><mo stretchy="false" form="postfix">}</mo><mi>l</mi><mi>o</mi><mi>g</mi><mi>L</mi><msub><mi>S</mi><mi>c</mi></msub></mrow><annotation encoding="application/x-tex">
log \mu_{c,g} = \varsigma_{g} + \{\alpha + h_g(x_c, y_c;\gamma_g)\}logLS_c
</annotation></semantics></math></p>
<p>where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Œº</mi><mrow><mi>c</mi><mo>,</mo><mi>g</mi></mrow></msub><annotation encoding="application/x-tex">\mu_{c,g}</annotation></semantics></math>
is the mean of gene
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>g</mi><annotation encoding="application/x-tex">g</annotation></semantics></math>
at location
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>c</mi><annotation encoding="application/x-tex">c</annotation></semantics></math>,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mo>ùúç</mo><mi>g</mi></msub><annotation encoding="application/x-tex">\varsigma_g</annotation></semantics></math>
is the log-mean of gene
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>g</mi><annotation encoding="application/x-tex">g</annotation></semantics></math>,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>h</mi><mi>g</mi></msub><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>x</mi><mi>c</mi></msub><mo>,</mo><msub><mi>y</mi><mi>c</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">h_g(x_c, y_c)</annotation></semantics></math>
is a smooth function of the spatial coordinates
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>x</mi><mi>c</mi></msub><annotation encoding="application/x-tex">x_c</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>y</mi><mi>c</mi></msub><annotation encoding="application/x-tex">y_c</annotation></semantics></math>
representing the library size effect, and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Œ±</mi><annotation encoding="application/x-tex">\alpha</annotation></semantics></math>
is the log-mean of the library size.</p>
<p>The <code><a href="../reference/SpaNormSVG.html">SpaNormSVG()</a></code> function fits this model and then uses
an F-test to identify spatially variable genes (SVGs).</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">HumanDLPFC</span> <span class="op">=</span> <span class="fu"><a href="../reference/SpaNormSVG.html">SpaNormSVG</a></span><span class="op">(</span><span class="va">HumanDLPFC</span><span class="op">)</span></span>
<span><span class="va">HumanDLPFC</span></span>
<span><span class="co">#&gt; class: SpatialExperiment </span></span>
<span><span class="co">#&gt; dim: 2508 4015 </span></span>
<span><span class="co">#&gt; metadata(1): SpaNorm</span></span>
<span><span class="co">#&gt; assays(2): counts logcounts</span></span>
<span><span class="co">#&gt; rownames(2508): ISG15 SDF4 ... MT-ND6 MT-CYB</span></span>
<span><span class="co">#&gt; rowData names(5): gene_name gene_biotype svg.F svg.p svg.fdr</span></span>
<span><span class="co">#&gt; colnames(4015): AAACAAGTATCTCCCA-1 AAACACCAATAACTGC-1 ...</span></span>
<span><span class="co">#&gt;   TTGTTTCCATACAACT-1 TTGTTTGTGTAAATTC-1</span></span>
<span><span class="co">#&gt; colData names(3): cell_count sample_id AnnotatedCluster</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : pxl_col_in_fullres pxl_row_in_fullres</span></span>
<span><span class="co">#&gt; imgData names(4): sample_id image_id data scaleFactor</span></span></code></pre></div>
<p>The <code><a href="../reference/topSVGs.html">topSVGs()</a></code> function can be used to retrieve the top
spatially variable genes (SVGs) at a given false discovery rate (FDR).
These are stored in the <code>rowData</code> slot of the
SpatialExperiment object.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">svgs</span> <span class="op">=</span> <span class="fu"><a href="../reference/topSVGs.html">topSVGs</a></span><span class="op">(</span><span class="va">HumanDLPFC</span>, n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">svgs</span></span>
<span><span class="co">#&gt;            svg.F         svg.p       svg.fdr gene_name   gene_biotype</span></span>
<span><span class="co">#&gt; SNAP25  55.12275 2.545221e-316 6.383415e-313    SNAP25 protein_coding</span></span>
<span><span class="co">#&gt; ITM2B   42.26465 4.588429e-248 5.753890e-245     ITM2B protein_coding</span></span>
<span><span class="co">#&gt; NEFL    38.63761 7.667892e-228 6.410358e-225      NEFL protein_coding</span></span>
<span><span class="co">#&gt; S100A11 34.78994 6.804583e-206 4.266474e-203   S100A11 protein_coding</span></span>
<span><span class="co">#&gt; MBP     25.55204 4.827547e-151 2.421497e-148       MBP protein_coding</span></span>
<span><span class="co">#&gt; STMN2   24.87049 7.123569e-147 2.977652e-144     STMN2 protein_coding</span></span>
<span><span class="co">#&gt; SAA1    22.60034 7.224814e-133 2.588548e-130      SAA1 protein_coding</span></span>
<span><span class="co">#&gt; SCGB1D2 22.21822 1.715386e-130 5.377736e-128   SCGB1D2 protein_coding</span></span>
<span><span class="co">#&gt; HPCAL1  21.93901 9.404551e-129 2.620735e-126    HPCAL1 protein_coding</span></span>
<span><span class="co">#&gt; TMSB10  21.20707 3.505727e-124 8.792363e-122    TMSB10 protein_coding</span></span></code></pre></div>
<p>We can visualise the spatially variable genes using the
<code><a href="../reference/plotSpatial.html">plotSpatial()</a></code> function.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">svgs</span><span class="op">$</span><span class="va">gene_name</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">9</span><span class="op">]</span>, <span class="kw">function</span><span class="op">(</span><span class="va">g</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="../reference/plotSpatial.html">plotSpatial</a></span><span class="op">(</span><span class="va">HumanDLPFC</span>, colour <span class="op">=</span> <span class="op">!</span><span class="op">!</span><span class="fu"><a href="https://rlang.r-lib.org/reference/sym.html" class="external-link">sym</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span>, what <span class="op">=</span> <span class="st">"expression"</span>, assay <span class="op">=</span> <span class="st">"logcounts"</span>, size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_colour_viridis_c</a></span><span class="op">(</span>option <span class="op">=</span> <span class="st">"F"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">wrap_plots</span><span class="op">(</span>ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p><img src="SpaNorm_files/figure-html/unnamed-chunk-15-1.png" width="1152"></p>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.4.2 (2024-10-31)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 24.04.1 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] SpatialExperiment_1.16.0    SingleCellExperiment_1.28.1</span></span>
<span><span class="co">#&gt;  [3] SummarizedExperiment_1.36.0 Biobase_2.66.0             </span></span>
<span><span class="co">#&gt;  [5] GenomicRanges_1.58.0        GenomeInfoDb_1.42.3        </span></span>
<span><span class="co">#&gt;  [7] IRanges_2.40.1              S4Vectors_0.44.0           </span></span>
<span><span class="co">#&gt;  [9] BiocGenerics_0.52.0         MatrixGenerics_1.18.1      </span></span>
<span><span class="co">#&gt; [11] matrixStats_1.5.0           patchwork_1.3.0            </span></span>
<span><span class="co">#&gt; [13] ggplot2_3.5.1               SpaNorm_1.1.2              </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] tidyselect_1.2.1        viridisLite_0.4.2       dplyr_1.1.4            </span></span>
<span><span class="co">#&gt;  [4] farver_2.1.2            fastmap_1.2.0           bluster_1.16.0         </span></span>
<span><span class="co">#&gt;  [7] digest_0.6.37           rsvd_1.0.5              lifecycle_1.0.4        </span></span>
<span><span class="co">#&gt; [10] cluster_2.1.6           statmod_1.5.0           magrittr_2.0.3         </span></span>
<span><span class="co">#&gt; [13] compiler_4.4.2          rlang_1.1.5             sass_0.4.9             </span></span>
<span><span class="co">#&gt; [16] tools_4.4.2             igraph_2.1.4            yaml_2.3.10            </span></span>
<span><span class="co">#&gt; [19] knitr_1.49              S4Arrays_1.6.0          labeling_0.4.3         </span></span>
<span><span class="co">#&gt; [22] dqrng_0.4.1             htmlwidgets_1.6.4       DelayedArray_0.32.0    </span></span>
<span><span class="co">#&gt; [25] RColorBrewer_1.1-3      abind_1.4-8             BiocParallel_1.40.0    </span></span>
<span><span class="co">#&gt; [28] withr_3.0.2             desc_1.4.3              grid_4.4.2             </span></span>
<span><span class="co">#&gt; [31] beachmat_2.22.0         colorspace_2.1-1        edgeR_4.4.2            </span></span>
<span><span class="co">#&gt; [34] scales_1.3.0            cli_3.6.3               rmarkdown_2.29         </span></span>
<span><span class="co">#&gt; [37] crayon_1.5.3            ragg_1.3.3              generics_0.1.3         </span></span>
<span><span class="co">#&gt; [40] metapod_1.14.0          httr_1.4.7              rjson_0.2.23           </span></span>
<span><span class="co">#&gt; [43] scuttle_1.16.0          cachem_1.1.0            zlibbioc_1.52.0        </span></span>
<span><span class="co">#&gt; [46] splines_4.4.2           parallel_4.4.2          BiocManager_1.30.25    </span></span>
<span><span class="co">#&gt; [49] XVector_0.46.0          vctrs_0.6.5             Matrix_1.7-1           </span></span>
<span><span class="co">#&gt; [52] jsonlite_1.8.9          BiocSingular_1.22.0     BiocNeighbors_2.0.1    </span></span>
<span><span class="co">#&gt; [55] irlba_2.3.5.1           systemfonts_1.2.1       magick_2.8.5           </span></span>
<span><span class="co">#&gt; [58] locfit_1.5-9.11         limma_3.62.2            jquerylib_0.1.4        </span></span>
<span><span class="co">#&gt; [61] glue_1.8.0              pkgdown_2.1.1           codetools_0.2-20       </span></span>
<span><span class="co">#&gt; [64] gtable_0.3.6            UCSC.utils_1.2.0        ScaledMatrix_1.14.0    </span></span>
<span><span class="co">#&gt; [67] munsell_0.5.1           tibble_3.2.1            pillar_1.10.1          </span></span>
<span><span class="co">#&gt; [70] htmltools_0.5.8.1       GenomeInfoDbData_1.2.13 R6_2.5.1               </span></span>
<span><span class="co">#&gt; [73] textshaping_1.0.0       evaluate_1.0.3          lattice_0.22-6         </span></span>
<span><span class="co">#&gt; [76] BiocStyle_2.34.0        scran_1.34.0            bslib_0.9.0            </span></span>
<span><span class="co">#&gt; [79] Rcpp_1.0.14             SparseArray_1.6.1       xfun_0.50              </span></span>
<span><span class="co">#&gt; [82] fs_1.6.5                prettydoc_0.4.1         pkgconfig_2.0.3</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-Atta2023" class="csl-entry">
Atta, Lyla, Kalen Clifton, Manjari Anant, Gohta Aihara, and Jean Fan.
2024. <span>‚ÄúGene Count Normalization in Single-Cell Imaging-Based
Spatially Resolved Transcriptomics.‚Äù</span> <em>bioRxiv</em>. <a href="https://doi.org/10.1101/2023.08.30.555624" class="external-link">https://doi.org/10.1101/2023.08.30.555624</a>.
</div>
<div id="ref-Bhuva2024" class="csl-entry">
Bhuva, Dharmesh D, Chin Wee Tan, Agus Salim, Claire Marceaux, Marie A
Pickering, Jinjin Chen, Malvika Kharbanda, et al. 2024. <span>‚ÄúLibrary
Size Confounds Biology in Spatial Transcriptomics Data.‚Äù</span>
<em>Genome Biology</em> 25 (1): 99.
</div>
<div id="ref-Maynard2021" class="csl-entry">
Maynard, Kristen R, Leonardo Collado-Torres, Lukas M Weber, Cedric
Uytingco, Brianna K Barry, Stephen R Williams, Joseph L Catallini, et
al. 2021. <span>‚ÄúTranscriptome-Scale Spatial Gene Expression in the
Human Dorsolateral Prefrontal Cortex.‚Äù</span> <em>Nature
Neuroscience</em> 24 (3): 425‚Äì36.
</div>
<div id="ref-Salim2022" class="csl-entry">
Salim, Agus, Ramyar Molania, Jianan Wang, Alysha De Livera, Rachel
Thijssen, and Terence P Speed. 2022. <span>‚Äú<span class="nocase">RUV-III-NB: normalization of single cell RNA-seq
data</span>.‚Äù</span> <em>Nucleic Acids Research</em> 50 (16): e96‚Äì96. <a href="https://doi.org/10.1093/nar/gkac486" class="external-link">https://doi.org/10.1093/nar/gkac486</a>.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Dharmesh D. Bhuva, Agus Salim, Ahmed Mohamed.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
